File: java/src/main/java/com/kcl/KCLResult.java
package com.kcl;

import java.util.List;
import java.util.Map;
import java.util.Optional;

import com.fasterxml.jackson.core.JsonProcessingException;

/**
 * Structured view of one configuration document produced by a KCL run. The
 * parsed value (a {@code Map}, {@code List} or scalar decoded from the
 * runtime JSON stream) plus the raw YAML document the runtime emitted for it.
 */
public final class KCLResult {
    private final Object result;
    private final String yamlDocument;

    KCLResult(Object result, String yamlDocument) {
        this.result = result;
        this.yamlDocument = yamlDocument;
    }

    /**
     * The raw decoded value of this document: a {@code Map<String, Object>},
     * {@code List<Object>} or scalar ({@code String}, {@code Number},
     * {@code Boolean}), or {@code null}.
     */
    public Object getValue() {
        return result;
    }

    /**
     * Look up {@code key} in the parsed value: dotted keys navigate nested
     * maps and numeric segments index into lists. Returns {@code null} when
     * the key is absent.
     */
    public Object get(String key) {
        Object value = result;
        for (String part : key.split("\\.")) {
            if (value instanceof Map) {
                value = ((Map<?, ?>) value).get(part);
            } else if (value instanceof List && isIndex(part)) {
                List<?> list = (List<?>) value;
                int index = Integer.parseInt(part);
                value = index >= 0 && index < list.size() ? list.get(index) : null;
            } else {
                return null;
            }
            if (value == null) {
                return null;
            }
        }
        return value;
    }

    /**
     * Typed variant of {@link #get(String)}: the value at {@code key}
     * converted to {@code target} (numeric widening allowed, everything else
     * strict). Empty when the key is absent.
     *
     * @throws KclException
     *             when the value exists but cannot be represented as
     *             {@code target}
     */
    public <T> Optional<T> get(String key, Class<T> target) {
        Object value = get(key);
        if (value == null) {
            return Optional.empty();
        }
        return Optional.of(convert(key, value, target));
    }

    /** The value at {@code key} as a string; empty when absent. */
    public Optional<String> getString(String key) {
        return get(key, String.class);
    }

    /** The value at {@code key} as an integer; empty when absent. */
    public Optional<Integer> getInt(String key) {
        return get(key, Integer.class);
    }

    /** The value at {@code key} as a double; empty when absent. */
    public Optional<Double> getDouble(String key) {
        return get(key, Double.class);
    }

    /** The value at {@code key} as a boolean; empty when absent. */
    public Optional<Boolean> getBoolean(String key) {
        return get(key, Boolean.class);
    }

    /** The whole document as a string, when it is one. */
    public Optional<String> asString() {
        return as(String.class);
    }

    /** The whole document as an integer, when it is numeric. */
    public Optional<Integer> asInt() {
        return as(Integer.class);
    }

    /** The whole document as a double, when it is numeric. */
    public Optional<Double> asDouble() {
        return as(Double.class);
    }

    /** The whole document as a boolean, when it is one. */
    public Optional<Boolean> asBoolean() {
        return as(Boolean.class);
    }

    /**
     * The whole document as a map.
     *
     * @throws KclException
     *             when the document is not a map
     */
    @SuppressWarnings("unchecked")
    public Map<String, Object> toMap() {
        if (!(result instanceof Map)) {
            throw new KclException("failed to convert result to map: type mismatch, got "
                    + (result == null ? "null" : result.getClass().getSimpleName()));
        }
        return (Map<String, Object>) result;
    }

    /**
     * The whole document as a list.
     *
     * @throws KclException
     *             when the document is not a list
     */
    @SuppressWarnings("unchecked")
    public List<Object> toList() {
        if (!(result instanceof List)) {
            throw new KclException("failed to convert result to list: type mismatch, got "
                    + (result == null ? "null" : result.getClass().getSimpleName()));
        }
        return (List<Object>) result;
    }

    /**
     * The raw YAML document emitted by the runtime for this result (one
     * slice of the multi-document {@code yaml_result}), or {@code ""} when
     * unavailable.
     */
    public String yamlString() {
        return yamlDocument == null ? "" : yamlDocument;
    }

    /** The document re-encoded as JSON. */
    public String jsonString() {
        try {
            return KclMappers.json().writerWithDefaultPrettyPrinter().writeValueAsString(result);
        } catch (JsonProcessingException e) {
            throw new IllegalStateException("failed to encode KCL result as JSON", e);
        }
    }

    private <T> Optional<T> as(Class<T> target) {
        if (result == null) {
            return Optional.empty();
        }
        return Optional.of(convert("<result>", result, target));
    }

    private static boolean isIndex(String part) {
        if (part.isEmpty() || part.length() > 9) {
            return false;
        }
        for (int i = 0; i < part.length(); i++) {
            if (!Character.isDigit(part.charAt(i))) {
                return false;
            }
        }
        return true;
    }

    private static <T> T convert(String key, Object value, Class<T> target) {
        if (target.isInstance(value)) {
            return target.cast(value);
        }
        if (target == Integer.class && value instanceof Number) {
            // numeric widening, mirroring kcl-go GetValue float64-to-*int
            return target.cast(Integer.valueOf(((Number) value).intValue()));
        }
        if (target == Double.class && value instanceof Number) {
            return target.cast(Double.valueOf(((Number) value).doubleValue()));
        }
        throw new KclException("failed to convert " + key + " to " + target.getSimpleName()
                + ": type mismatch, got " + (value == null ? "null" : value.getClass().getSimpleName()));
    }
}

File: kotlin/src/main/kotlin/com/kcl/KCLResult.kt
package com.kcl

import com.fasterxml.jackson.core.JsonProcessingException

/**
 * Structured view of one configuration document produced by a KCL run:
 * the parsed value (a [Map], [List] or scalar decoded from the runtime JSON
 * stream) plus the raw YAML document the runtime emitted for it.
 */
data class KCLResult internal constructor(
    /** The raw decoded value: a `Map<String, Any?>`, `List<Any?>` or scalar,
     * or `null`. */
    val value: Any?,
    /** The raw YAML document emitted by the runtime for this result (one
     * slice of the multi-document `yaml_result`), or `null` when
     * unavailable. */
    val yamlDocument: String?,
) {

    /**
     * Look up [key] in the parsed value, mirroring kcl-go's
     * `KCLResult.Get`. Dotted keys navigate nested maps and numeric segments
     * index into lists. Returns `null` when the key is absent.
     */
    fun get(key: String): Any? {
        var current: Any? = value
        for (part in key.split('.')) {
            current = when (current) {
                is Map<*, *> -> current[part]
                is List<*> -> {
                    val index = part.toIntOrNull()
                    if (index != null && index >= 0 && index < current.size) current[index] else null
                }
                else -> return null
            }
            if (current == null) {
                return null
            }
        }
        return current
    }

    /**
     * Typed variant of [get], mirroring kcl-go's `Get(key, target)`:
     * coerces the value at [key] to [target], throwing [KclException] on a
     * type mismatch (numeric widening is allowed, everything else is strict).
     * Returns `null` when the key is absent.
     */
    @Suppress("UNCHECKED_CAST")
    fun <T> get(key: String, target: Class<T>): T? {
        val value = get(key) ?: return null
        if (target.isInstance(value)) {
            return target.cast(value)
        }
        // Accept both the wrapper and the primitive class form (Kotlin's
        // `Int::class.java` is the primitive int.class).
        if ((target == Integer::class.javaObjectType || target == Integer::class.java) && value is Number) {
            return value.toInt() as T
        }
        if ((target == Double::class.javaObjectType || target == Double::class.java) && value is Number) {
            return value.toDouble() as T
        }
        throw KclException(
            "failed to convert $key to ${target.simpleName}: type mismatch, got " +
                (value?.javaClass?.simpleName ?: "null")
        )
    }

    /** Reified shorthand for `get(key, T::class.java)`. */
    fun <T : Any> get(key: String, target: kotlin.reflect.KClass<T>): T? = get(key, target.java)

    /** The value at [key] as a [String]; `null` when absent. */
    fun getString(key: String): String? {
        val value = get(key) ?: return null
        return value as? String
            ?: throw KclException("failed to convert $key to String: type mismatch")
    }

    /** The value at [key] as an [Int]; `null` when absent. */
    fun getInt(key: String): Int? {
        val value = get(key) ?: return null
        return (value as? Number)?.toInt()
            ?: throw KclException("failed to convert $key to Int: type mismatch")
    }

    /** The value at [key] as a [Double]; `null` when absent. */
    fun getDouble(key: String): Double? {
        val value = get(key) ?: return null
        return (value as? Number)?.toDouble()
            ?: throw KclException("failed to convert $key to Double: type mismatch")
    }

    /** The value at [key] as a [Boolean]; `null` when absent. */
    fun getBoolean(key: String): Boolean? {
        val value = get(key) ?: return null
        return value as? Boolean
            ?: throw KclException("failed to convert $key to Boolean: type mismatch")
    }

    /** The whole document as a map, mirroring kcl-go's `ToMap`. */
    @Suppress("UNCHECKED_CAST")
    fun toMap(): Map<String, Any?> = value as? Map<String, Any?>
        ?: throw KclException(
            "failed to convert result to map: type mismatch, got " +
                (value?.javaClass?.simpleName ?: "null")
        )

    /** The whole document as a list, mirroring kcl-go's `ToList`. */
    @Suppress("UNCHECKED_CAST")
    fun toList(): List<Any?> = value as? List<Any?>
        ?: throw KclException(
            "failed to convert result to list: type mismatch, got " +
                (value?.javaClass?.simpleName ?: "null")
        )

    /**
     * The raw YAML document emitted by the runtime for this result (one
     * slice of the multi-document `yaml_result`), or an empty string when
     * unavailable.
     */
    fun yamlString(): String = yamlDocument ?: ""

    /** The document re-encoded as JSON, mirroring kcl-go's `JSONString`. */
    fun jsonString(): String = try {
        KclMappers.json.writerWithDefaultPrettyPrinter().writeValueAsString(value)
    } catch (e: JsonProcessingException) {
        throw IllegalStateException("failed to encode KCL result as JSON", e)
    }
}

/**
 * Collection of [KCLResult] documents produced by a run, mirroring kcl-go's
 * `KCLResultList`. The raw runtime outputs are kept for callers that need the
 * untouched response.
 */
class KCLResultList internal constructor(
    private val results: List<KCLResult>,
    /** The untouched `json_result` emitted by the runtime. */
    val rawJsonResult: String,
    /** The untouched `yaml_result` emitted by the runtime. */
    val rawYamlResult: String,
) : List<KCLResult> by results {

    /** The first document, or `null` when empty. */
    fun first(): KCLResult? = results.firstOrNull()

    /** The last document, or `null` when empty. */
    fun tail(): KCLResult? = results.lastOrNull()

    /** The first document as a map, mirroring kcl-go's `KCLResultList.ToMap`. */
    fun toMap(): Map<String, Any?> =
        firstOrNull()?.toMap() ?: throw KclException("result is nil")

    /** Dotted-key lookup on the first document; `null` when absent. */
    fun get(key: String): Any? = firstOrNull()?.get(key)
}
